packages/notificator/notificator.go

```diff
@@ -2,122 +2,131 @@ package notificator
-				return false
-			}
-
-			id, err := strconv.ParseInt(notif["id"], 10, 64)
-			if err != nil {
-				log.WithFields(log.Fields{"type": consts.ConversionError, "error": err}).Error("conversion string to int64")
-				return false
-			}
-
-			lni, ok := notifications.Load(ecosystemID)
-			ln := lni.(NotificationStats)
-			if ok && *ln.lastNotifID < id {
-				*ln.lastNotifID = id
-				notifications.Store(ecosystemID, ln)
+				continue
 			}
 		}
-
-		return true
-	})
+	}
 }
 
-func mapToString(value map[string]string) (string, error) {
-	bytes, err := json.Marshal(value)
+func getEcosystemIDList() ([]int64, error) {
+	var idlist []int64
+
+	db := model.GetDB(nil)
+	rows, err := db.Raw("SELECT id FROM system_states").Rows()
+	defer rows.Close()
+
 	if err != nil {
-		return "", err
+		return nil, err
 	}
-	return string(bytes), nil
+
+	for rows.Next() {
+		var id int64
+		rows.Scan(&id)
+		idlist = append(idlist, id)
+	}
+
+	return idlist, err
 }
 
-func getEcosystemNotifications(ecosystemID EcosystemID, lastNotificationID int64, userIDs NotificationStats) []map[string]string {
-	users := make([]int64, 0)
-	userIDs.userIDs.Range(func(key, value interface{}) bool {
-		users = append(users, int64(key.(UserID)))
-		return true
-	})
+func parseRecipientNotification(rows []map[string]string) (map[int64]*[]notificationRecord, error) {
+	recipientNotifications := make(map[int64]*[]notificationRecord)
 
-	rows, err := model.GetAllNotifications(int64(ecosystemID), lastNotificationID, users)
-	if err != nil || len(rows) == 0 {
+	convert := func(dataRow map[string]string, value string, errMessage string) (int64, error) {
+		var result int64
+		result, err := strconv.ParseInt(dataRow[value], 10, 64)
 		if err != nil {
-			log.WithFields(log.Fields{"type": consts.DBError, "error": err}).Error("getting all notifications")
+			log.WithFields(log.Fields{"type": consts.ConversionError, "error": err}).Error(errMessage)
 		}
-		return nil
+		return result, err
 	}
-	return rows
-}
 
-// AddUser is subscribing user to notifications
-func AddUser(userID int64, ecosystemID int64) {
-	eId := EcosystemID(ecosystemID)
+	for _, r := range rows {
+		recipientId, err := convert(r, "recipient_id", "error converting records count")
+		if err != nil {
+			return recipientNotifications, err
+		}
 
-	var ns NotificationStats
-	ins, ok := notifications.Load(eId)
+		roleId, err := convert(r, "role_id", "error converting records count")
+		if err != nil {
+			return recipientNotifications, err
+		}
+
+		count, err := convert(r, "cnt", "error converting records count")
+		if err != nil {
+			return recipientNotifications, err
+		}
+
+		roleNotifications := notificationRecord{
+			RoleID:       roleId,
+			RecordsCount: count,
+		}
+
+		nr, ok := recipientNotifications[recipientId]
+		if ok {
+			*nr = append(*nr, roleNotifications)
+			continue
+		}
+
+		records := []notificationRecord{
+			roleNotifications,
```

Comment:
why you use an array with only one item? May be array here is unnecessary?

Reply:
values can be added to the next iterations